server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -201,7 +193,15 @@ public void handleException(final TransportException exp) {
             }
         }
 
-        private void retry(final Throwable failure, final Predicate<ClusterState> statePredicate) {
+        private void retryOnMasterChange(ClusterState state, Throwable failure) {
+            retry(state, failure, MasterNodeChangePredicate.build(state));
+        }
+
+        private void retry(ClusterState state, final Throwable failure, final Predicate<ClusterState> statePredicate) {
+            if (observer == null) {
+                this.observer =
+                    new ClusterStateObserver(state, clusterService, request.masterNodeTimeout(), logger, threadPool.getThreadContext());
```

Comment:
This change means that the timeout starts counting from the first retry, rather than from the start of execution of the action. I think we should adjust the timeout to account for this since this retry might be happening a long time later.

Reply:
Fair point, I pushed 29806e8 :)